src/handshake/client.rs

```diff
@@ -96,17 +96,22 @@ impl<S: Read + Write> HandshakeRole for ClientHandshake<S> {
 fn generate_request(request: Request, key: &str) -> Result<Vec<u8>> {
     let mut req = Vec::new();
     let uri = request.uri();
+    let headers = request.headers();
 
-    let authority = uri.authority().ok_or(Error::Url(UrlError::NoHostName))?.as_str();
-    let host = if let Some(idx) = authority.find('@') {
-        // handle possible name:password@
-        authority.split_at(idx + 1).1
+    let host = if headers.contains_key(header::HOST) {
+        headers.get(header::HOST).unwrap().to_str()?
```

Comment:
I think this can be rewritten as:

```rust
if let Some(host) = headers.get(header::HOST) {
}
```

Then we don't need to `unwrap()` additionally.

Reply:
But the `host` will not live long enough for `write!` use, it's only available in `if` block

Reply:
Won't it? I think its lifetime should be identical in both cases (as it is the same as `headers`), but I might be wrong.